File: examples/c/example_08_site_dimension.c
#include "cmip7_c_common.h"

#include <stdio.h>

#define NSITES 126
#define NVERTICES 2

int main(int argc, char **argv) {
  const char *repo_root = argc > 1 ? argv[1] : ".";
  const char *output_dir = argc > 2 ? argv[2] : "examples/c/output";
  char input_path[CMIP7_PATH_MAX];
  char tables_path[CMIP7_PATH_MAX];
  char cell_measures[CMOR_MAX_STRING];
  char long_name[CMOR_MAX_STRING];
  char filename[CMOR_MAX_STRING];
  int file_action = CMOR_REPLACE;
  int exit_control = CMOR_EXIT_ON_MAJOR;
  int table_id, grid_table_id, time_id, site_id, height_id, grid_id, var_id;
  int grid_axes[1];
  int axes[3];
  int site, time_index;
  double time[CMIP7_NTIMES] = {15.5, 45.5};
  double time_bnds[CMIP7_NTIMES + 1] = {0.0, 31.0, 60.0};
  double site_values[NSITES];
  double latitude[NSITES];
  double longitude[NSITES];
  double latitude_vertices[NSITES * NVERTICES];
  double longitude_vertices[NSITES * NVERTICES];
  double height = 2.0;
  float tas[CMIP7_NTIMES * NSITES];
  float missing = CMIP7_MISSING_VALUE;

  snprintf(tables_path, sizeof(tables_path), "%s/cmip7-cmor-tables/tables",
           repo_root);
  cmip7_write_user_input_json(output_dir, "example_08_input.json", "mon", "r1",
                              "f1", input_path, sizeof(input_path));

  cmor_setup(tables_path, &file_action, NULL, &exit_control, NULL, NULL);
  cmor_dataset_json(input_path);
  cmor_load_table("CMIP7_atmos.json", &table_id);

  cmor_axis(&time_id, "time1", "days since 1979-01-01", CMIP7_NTIMES, time,
            'd', time_bnds, 1, NULL);
  cmor_axis(&height_id, "height2m", "m", 1, &height, 'd', NULL, 0, NULL);

  for (site = 0; site < NSITES; ++site) {
    int vertex = site * NVERTICES;
    site_values[site] = (double)(site + 1);
    latitude[site] = -60.0 + 120.0 * (double)site / (double)(NSITES - 1);
    longitude[site] = 0.5 + 359.0 * (double)site / (double)(NSITES - 1);
    latitude_vertices[vertex] = latitude[site] - 0.25;
    latitude_vertices[vertex + 1] = latitude[site] + 0.25;
    longitude_vertices[vertex] = longitude[site] - 0.25;
    longitude_vertices[vertex + 1] = longitude[site] + 0.25;
  }
  cmor_axis(&site_id, "site", "1", NSITES, site_values, 'd', NULL, 0, NULL);

  cmor_load_table("CMIP7_grids.json", &grid_table_id);
  cmor_set_table(grid_table_id);
  grid_axes[0] = site_id;
  cmor_grid(&grid_id, 1, grid_axes, 'd', latitude, longitude, NVERTICES,
            latitude_vertices, longitude_vertices);

  cmor_load_table("CMIP7_atmos.json", &table_id);
  axes[0] = height_id;
  axes[1] = time_id;
  axes[2] = grid_id;
  cmor_variable(&var_id, "tas_tpt-h2m-hs-u", "K", 3, axes, 'f', &missing,
                NULL, NULL, NULL, NULL, NULL);

  cmip7_get_cell_measures(tables_path, "atmos", "tas_tpt-h2m-hs-u", "mon",
                          "glb", cell_measures, sizeof(cell_measures));
  cmor_set_variable_attribute(var_id, "cell_measures", 'c', cell_measures);
  if (cmip7_get_long_name_override(tables_path, "atmos", "tas_tpt-h2m-hs-u",
                                   "mon", "glb", long_name,
                                   sizeof(long_name))) {
    cmor_set_variable_attribute(var_id, "long_name", 'c', long_name);
  }

  for (time_index = 0; time_index < CMIP7_NTIMES; ++time_index) {
    for (site = 0; site < NSITES; ++site) {
      tas[time_index * NSITES + site] =
          275.0f + 15.0f * (float)(time_index * NSITES + site) /
                       (float)(CMIP7_NTIMES * NSITES - 1);
    }
  }
  cmor_write(var_id, tas, 'f', NULL, CMIP7_NTIMES, NULL, NULL, NULL);
  filename[0] = '\0';
  cmor_close_variable(var_id, filename, NULL);
  printf("%s\n", filename);
  cmor_close();
  return 0;
}

File: examples/fortran/example_08_site_dimension.f90
program example_08_site_dimension
  use cmip7_fortran_common
  implicit none

  integer, parameter :: nsites = 126
  integer, parameter :: nvertices = 2
  character(len=1024) :: repo_root
  character(len=1024) :: output_dir
  character(len=2048) :: tables_path
  character(len=2048) :: input_path
  character(len=2048) :: filename
  integer :: time_id
  integer :: site_id
  integer :: height_id
  integer :: grid_id
  integer :: grid_table_id
  integer :: var_id
  integer :: ierr
  integer :: nul_pos
  integer :: site
  integer :: time_index
  real :: tas(nsites, ntimes)
  double precision :: time(ntimes)
  double precision :: time_bnds(2, ntimes)
  double precision :: site_values(nsites)
  double precision :: latitude(nsites)
  double precision :: longitude(nsites)
  double precision :: latitude_vertices(nvertices, nsites)
  double precision :: longitude_vertices(nvertices, nsites)

  call get_example_args(repo_root, output_dir)
  call prepare_cmor_paths(repo_root, output_dir, "example_08_input.json", &
       "mon", "r1", "f1", tables_path, input_path)

  ierr = cmor_setup(inpath=trim(tables_path), &
       netcdf_file_action=CMOR_REPLACE, &
       exit_control=CMOR_EXIT_ON_MAJOR)
  call check_status("cmor_setup", ierr)
  ierr = cmor_dataset_json(trim(input_path))
  call check_status("cmor_dataset_json", ierr)

  ierr = cmor_load_table("CMIP7_atmos.json")
  call check_id("cmor_load_table(CMIP7_atmos)", ierr)
  time = (/ 15.5d0, 45.5d0 /)
  time_bnds(:, 1) = (/ 0.0d0, 31.0d0 /)
  time_bnds(:, 2) = (/ 31.0d0, 60.0d0 /)
  time_id = cmor_axis(table_entry="time1", &
       units="days since 1979-01-01", &
       length=ntimes, coord_vals=time, cell_bounds=time_bnds)
  call check_id("cmor_axis(time1)", time_id)
  height_id = cmor_axis(table_entry="height2m", units="m", length=1, &
       coord_vals=(/ 2.0d0 /))
  call check_id("cmor_axis(height2m)", height_id)

  do site = 1, nsites
    site_values(site) = dble(site)
    latitude(site) = -60.0d0 + 120.0d0 * dble(site - 1) / dble(nsites - 1)
    longitude(site) = 0.5d0 + 359.0d0 * dble(site - 1) / dble(nsites - 1)
    latitude_vertices(:, site) = (/ latitude(site) - 0.25d0, &
         latitude(site) + 0.25d0 /)
    longitude_vertices(:, site) = (/ longitude(site) - 0.25d0, &
         longitude(site) + 0.25d0 /)
  enddo
  site_id = cmor_axis(table_entry="site", units="1", length=nsites, &
       coord_vals=site_values)
  call check_id("cmor_axis(site)", site_id)
  grid_table_id = cmor_load_table("CMIP7_grids.json")
  call check_id("cmor_load_table(CMIP7_grids)", grid_table_id)
  call cmor_set_table(grid_table_id)
  grid_id = cmor_grid(axis_ids=(/ site_id /), latitude=latitude, &
       longitude=longitude, latitude_vertices=latitude_vertices, &
       longitude_vertices=longitude_vertices)
  call check_grid_id("cmor_grid", grid_id)

  ierr = cmor_load_table("CMIP7_atmos.json")
  call check_id("cmor_load_table(CMIP7_atmos)", ierr)
  var_id = cmor_variable(table_entry="tas_tpt-h2m-hs-u", units="K", &
       axis_ids=(/ grid_id, time_id /), missing_value=missing_value)
  call check_id("cmor_variable(tas)", var_id)
  call apply_cmip7_variable_metadata(var_id, tables_path, "atmos", &
       "tas_tpt-h2m-hs-u", "mon", "glb")

  do time_index = 1, ntimes
    do site = 1, nsites
      tas(site, time_index) = 275.0 + 15.0 * &
           real((time_index - 1) * nsites + site - 1) / real(ntimes * nsites - 1)
    enddo
  enddo
  ierr = cmor_write(var_id, tas)
  call check_status("cmor_write", ierr)

  filename = ""
  ierr = cmor_close(var_id, file_name=filename)
  call check_status("cmor_close(var)", ierr)
  nul_pos = index(filename, char(0))
  if (nul_pos > 0) filename(nul_pos:) = " "
  write(*, '(a)') trim(filename)
  ierr = cmor_close()
  call check_status("cmor_close", ierr)
end program example_08_site_dimension
